File: frontend/types/integration.ts
// Integration Types
// AUTO-SYNCED with backend via openapi-typescript
// Last synced: 2026-01-08
// Source: components["schemas"]["IntegrationResponse"], etc.

// ============================================
// ENUMS / UNION TYPES
// ============================================

export type EcommercePlatform = 'shopify' | 'woocommerce';

export type IntegrationStatus = 'active' | 'error' | 'paused' | 'disconnected';

export type SyncType = 'full' | 'incremental';

export type SyncStatus = 'idle' | 'syncing' | 'error';

export type HealthStatus = 'healthy' | 'unhealthy' | 'rate_limited' | 'unauthorized';

// ============================================
// OAUTH TYPES
// ============================================

/**
 * OAuth init request
 * Matches: components["schemas"]["OAuthInitRequest"]
 */
export interface OAuthInitRequest {
  platform: EcommercePlatform;
  store_url: string;
}

/**
 * OAuth init response
 * Matches: components["schemas"]["OAuthInitResponse"]
 */
export interface OAuthInitResponse {
  authorization_url: string;
  state: string;
}

// ============================================
// WOOCOMMERCE CONNECT
// ============================================

/**
 * WooCommerce connect request
 * Matches: components["schemas"]["WooCommerceConnectRequest"]
 */
export interface WooCommerceConnectRequest {
  store_url: string;
  store_name?: string | null;
  consumer_key: string;
  consumer_secret: string;
}

// ============================================
// INTEGRATION CRUD
// ============================================

/**
 * Integration response from GET endpoints
 * Matches: components["schemas"]["IntegrationResponse"]
 */
export interface Integration {
  id: string;
  user_id: string;
  platform: EcommercePlatform;
  store_url: string;
  store_name: string | null;
  status: IntegrationStatus;
  error_message: string | null;
  scopes: string[];
  last_sync_at: string | null;
  sync_status: SyncStatus;
  products_synced: number;
  settings: Record<string, unknown>;
  created_at: string;
  updated_at: string;
}

/**
 * Integration list response
 * Matches: components["schemas"]["IntegrationListResponse"]
 */
export interface IntegrationListResponse {
  integrations: Integration[];
  total: number;
}

/**
 * Integration update request
 * Matches: components["schemas"]["IntegrationUpdate"]
 */
export interface IntegrationUpdate {
  store_name?: string | null;
  status?: IntegrationStatus | null;
  settings?: Record<string, unknown> | null;
}

// ============================================
// SYNC TYPES
// ============================================

/**
 * Sync trigger request
 * Matches: components["schemas"]["SyncTriggerRequest"]
 */
export interface SyncTriggerRequest {
  sync_type?: SyncType;  // Default: "incremental"
}

/**
 * Sync status response
 * Matches: components["schemas"]["SyncStatusResponse"]
 */
export interface SyncStatusResponse {
  integration_id: string;
  sync_status: SyncStatus;
  last_sync_at: string | null;
  products_synced: number;
  current_progress?: number | null;
}

/**
 * Sync log entry
 * Matches: components["schemas"]["SyncLogResponse"]
 */
export interface SyncLog {
  id: string;
  integration_id: string;
  sync_type: SyncType;
  started_at: string;
  completed_at: string | null;
  duration_seconds: number | null;
  success: boolean;
  products_created: number;
  products_updated: number;
  products_deleted: number;
  error_details: string | null;
}

/**
 * Sync logs list response
 */
export interface SyncLogsResponse {
  items: SyncLog[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
}

// ============================================
// PRODUCT LINK TYPES
// ============================================

/**
 * Create product link request
 * Matches: components["schemas"]["ProductLinkCreate"]
 */
export interface ProductLinkCreate {
  product_id: string;
  external_product_id: string;
  external_variant_id?: string | null;
}

/**
 * Product link response
 * Matches: components["schemas"]["ProductLinkResponse"]
 */
export interface ProductLink {
  id: string;
  product_id: string;
  integration_id: string;
  external_product_id: string;
  external_variant_id: string | null;
  external_price: number | null;
  external_compare_at_price: number | null;
  last_price_push_at: string | null;
  last_price_pull_at: string | null;
  sync_enabled: boolean;
  created_at: string;
}

/**
 * Product link list response
 * Matches: components["schemas"]["ProductLinkListResponse"]
 */
export interface ProductLinkListResponse {
  links: ProductLink[];
  total: number;
}

// ============================================
// PRICE PUSH TYPES
// ============================================

/**
 * Price push request
 * Matches: components["schemas"]["PricePushRequest"]
 */
export interface PricePushRequest {
  product_link_id: string;
  new_price: number | string;
  compare_at_price?: number | string | null;
}

/**
 * Price push response
 * Matches: components["schemas"]["PricePushResponse"]
 */
export interface PricePushResponse {
  success: boolean;
  product_link_id: string;
  old_price: number | null;
  new_price: number;
  error: string | null;
}

/**
 * Bulk price push request
 * Matches: components["schemas"]["BulkPricePushRequest"]
 */
export interface BulkPricePushRequest {
  updates: PricePushRequest[];
}

/**
 * Bulk price push response
 * Matches: components["schemas"]["BulkPricePushResponse"]
 */
export interface BulkPricePushResponse {
  results: PricePushResponse[];
  success_count: number;
  failure_count: number;
}

// ============================================
// HEALTH CHECK TYPES
// ============================================

/**
 * Integration health response
 * Matches: components["schemas"]["IntegrationHealthResponse"]
 */
export interface IntegrationHealthResponse {
  integration_id: string;
  platform: EcommercePlatform;
  store_url: string;
  status: HealthStatus;
  checked_at: string;
}

// ============================================
// UI HELPER TYPES (Frontend-only)
// ============================================

export interface PlatformConfig {
  id: EcommercePlatform;
  name: string;
  logo: string;
  description: string;
  authType: 'oauth' | 'api_key';
  docsUrl: string;
}

export const PLATFORM_CONFIGS: Record<EcommercePlatform, PlatformConfig> = {
  shopify: {
    id: 'shopify',
    name: 'Shopify',
    logo: '/logos/shopify.svg',
    description: 'Connect your Shopify store via OAuth',
    authType: 'oauth',
    docsUrl: 'https://shopify.dev/docs/apps/auth/oauth',
  },
  woocommerce: {
    id: 'woocommerce',
    name: 'WooCommerce',
    logo: '/logos/woocommerce.svg',
    description: 'Connect with REST API keys',
    authType: 'api_key',
    docsUrl: 'https://woocommerce.com/document/woocommerce-rest-api/',
  },
};

File: backend/schemas/integration.py
# backend/schemas/integration.py

"""
Integration Schemas

Request/Response DTOs for e-commerce integration endpoints.

FIX (2026-01-24): Added proper Dict[str, Any] type annotations to fix Pylance warnings.
FIX (2026-01-27): Added consumer_key/consumer_secret to IntegrationUpdate for credential updates.
"""

from datetime import datetime
from decimal import Decimal
from enum import StrEnum
from typing import Any
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, field_validator

# ==================== Enums ====================


class EcommercePlatform(StrEnum):
    SHOPIFY = "shopify"
    WOOCOMMERCE = "woocommerce"


class IntegrationStatus(StrEnum):
    ACTIVE = "active"
    ERROR = "error"
    PAUSED = "paused"
    DISCONNECTED = "disconnected"


# ==================== OAuth Schemas ====================


class OAuthInitRequest(BaseModel):
    """Request to start OAuth flow"""

    platform: EcommercePlatform
    store_url: str = Field(..., min_length=3, max_length=255)

    @field_validator("store_url")
    @classmethod
    def validate_store_url(cls, v: str) -> str:
        v = v.strip().rstrip("/")
        # Preserve the protocol, just lowercase the domain
        if v.startswith("http://"):
            return "http://" + v[7:].lower()
        elif v.startswith("https://"):
            return "https://" + v[8:].lower()
        # No protocol provided - just lowercase
        return v.lower()


class OAuthInitResponse(BaseModel):
    """Response with OAuth authorization URL"""

    authorization_url: str
    state: str  # For CSRF verification on callback


class OAuthCallbackRequest(BaseModel):
    """OAuth callback data"""

    code: str
    state: str
    shop: str | None = None  # Shopify includes this


class WooCommerceConnectRequest(BaseModel):
    """Connect WooCommerce store with API keys"""

    store_url: str = Field(..., min_length=3, max_length=255)
    store_name: str | None = Field(None, max_length=255)
    consumer_key: str = Field(..., min_length=10)
    consumer_secret: str = Field(..., min_length=10)

    @field_validator("store_url")
    @classmethod
    def validate_store_url(cls, v: str) -> str:
        v = v.strip().rstrip("/")
        # Preserve the protocol, just lowercase the domain
        if v.startswith("http://"):
            return "http://" + v[7:].lower()
        elif v.startswith("https://"):
            return "https://" + v[8:].lower()
        # No protocol provided - just lowercase
        return v.lower()

    @field_validator("consumer_key")
    @classmethod
    def validate_consumer_key(cls, v: str) -> str:
        if not v.startswith("ck_"):
            raise ValueError("Consumer key must start with 'ck_'")
        return v

    @field_validator("consumer_secret")
    @classmethod
    def validate_consumer_secret(cls, v: str) -> str:
        if not v.startswith("cs_"):
            raise ValueError("Consumer secret must start with 'cs_'")
        return v


# ==================== Integration CRUD Schemas ====================


class IntegrationCreate(BaseModel):
    """Manual integration creation (for WooCommerce API keys)"""

    platform: EcommercePlatform
    store_url: str = Field(..., min_length=3, max_length=255)
    store_name: str | None = Field(None, max_length=255)
    # For WooCommerce manual setup
    consumer_key: str | None = None
    consumer_secret: str | None = None


class IntegrationUpdate(BaseModel):
    """
    Update integration settings.

    FIX (2026-01-27): Added consumer_key and consumer_secret fields to allow
    updating WooCommerce credentials without deleting/reconnecting the integration.
    """

    store_name: str | None = Field(None, max_length=255)
    status: IntegrationStatus | None = None
    settings: dict[str, Any] | None = None
    # NEW (2026-01-27): Allow credential updates for WooCommerce reconnection
    consumer_key: str | None = Field(None, min_length=10)
    consumer_secret: str | None = Field(None, min_length=10)

    @field_validator("consumer_key")
    @classmethod
    def validate_consumer_key(cls, v: str | None) -> str | None:
        if v is not None and not v.startswith("ck_"):
            raise ValueError("Consumer key must start with 'ck_'")
        return v

    @field_validator("consumer_secret")
    @classmethod
    def validate_consumer_secret(cls, v: str | None) -> str | None:
        if v is not None and not v.startswith("cs_"):
            raise ValueError("Consumer secret must start with 'cs_'")
        return v


class IntegrationResponse(BaseModel):
    """Integration response (public fields only - no secrets)"""

    model_config = ConfigDict(from_attributes=True)
    id: UUID
    platform: EcommercePlatform
    store_url: str
    store_name: str | None
    status: IntegrationStatus
    error_message: str | None
    scopes: list[str]
    last_sync_at: datetime | None
    sync_status: str
    products_synced: int
    settings: dict[str, Any]
    created_at: datetime
    updated_at: datetime


class IntegrationListResponse(BaseModel):
    """List of integrations"""

    integrations: list[IntegrationResponse]
    total: int


# ==================== Sync Schemas ====================


class SyncTriggerRequest(BaseModel):
    """Request to trigger a product sync"""

    sync_type: str = Field(default="full", pattern="^(full|incremental)$")


class SyncStatusResponse(BaseModel):
    """Current sync status"""

    integration_id: UUID
    sync_status: str  # idle, syncing, error
    last_sync_at: datetime | None
    products_synced: int
    current_progress: int | None = None  # For ongoing syncs


class SyncLogResponse(BaseModel):
    """Sync log entry"""

    model_config = ConfigDict(from_attributes=True)
    id: UUID
    sync_type: str
    started_at: datetime
    completed_at: datetime | None
    duration_seconds: float | None
    success: bool
    products_created: int
    products_updated: int
    products_deleted: int
    error_details: str | None


class SyncLogsListResponse(BaseModel):
    """List of sync logs"""

    logs: list[SyncLogResponse]
    total: int


# ==================== Product Link Schemas ====================


class ProductLinkCreate(BaseModel):
    """Link an SSP product to an external platform product"""

    product_id: UUID
    external_product_id: str = Field(..., max_length=100)
    external_variant_id: str | None = Field(None, max_length=100)


class ProductLinkResponse(BaseModel):
    """Product link response"""

    model_config = ConfigDict(from_attributes=True)
    id: UUID
    product_id: UUID
    integration_id: UUID
    external_product_id: str
    external_variant_id: str | None
    external_price: float | None
    external_compare_at_price: float | None
    last_price_push_at: datetime | None
    last_price_pull_at: datetime | None
    sync_enabled: bool
    created_at: datetime


class ProductLinkListResponse(BaseModel):
    """List of product links"""

    links: list[ProductLinkResponse]
    total: int


# ==================== Price Push Schemas ====================


class PricePushRequest(BaseModel):
    """Request to push a price update to the platform"""

    product_link_id: UUID
    new_price: Decimal = Field(..., gt=0, decimal_places=2)
    compare_at_price: Decimal | None = Field(None, gt=0, decimal_places=2)


class PricePushResponse(BaseModel):
    """Result of price push operation"""

    success: bool
    product_link_id: UUID
    old_price: Decimal | None
    new_price: Decimal
    error: str | None = None


class BulkPricePushRequest(BaseModel):
    """Push multiple price updates"""

    updates: list[PricePushRequest] = Field(..., min_length=1, max_length=100)


class BulkPricePushResponse(BaseModel):
    """Results of bulk price push"""

    results: list[PricePushResponse]
    success_count: int
    failure_count: int


# ==================== Webhook Schemas ====================


class WebhookPayload(BaseModel):
    """Generic webhook payload wrapper"""

    topic: str
    shop: str
    payload: dict[str, Any]


# ==================== Health Check ====================


class IntegrationHealthResponse(BaseModel):
    """Health check response for an integration"""

    integration_id: UUID
    platform: EcommercePlatform
    store_url: str
    status: str  # healthy, unhealthy, rate_limited, unauthorized
    checked_at: datetime
